hotshot-types/src/vid/advz.rs

```diff
@@ -0,0 +1,259 @@
+impl VidScheme for ADVZScheme {
+    type Commit = <Advz as VidScheme>::Commit;
+    type Share = <Advz as VidScheme>::Share;
+    type Common = <Advz as VidScheme>::Common;
+
+    fn commit_only<B>(&mut self, payload: B) -> VidResult<Self::Commit>
+    where
+        B: AsRef<[u8]>,
+    {
+        self.0.commit_only(payload)
+    }
+
+    fn disperse<B>(&mut self, payload: B) -> VidResult<VidDisperse<Self>>
+    where
+        B: AsRef<[u8]>,
+    {
+        self.0.disperse(payload).map(vid_disperse_conversion)
+    }
+
+    fn verify_share(
+        &self,
+        share: &Self::Share,
+        common: &Self::Common,
+        commit: &Self::Commit,
+    ) -> VidResult<Result<(), ()>> {
+        self.0.verify_share(share, common, commit)
+    }
+
+    fn recover_payload(&self, shares: &[Self::Share], common: &Self::Common) -> VidResult<Vec<u8>> {
+        self.0.recover_payload(shares, common)
+    }
+
+    fn is_consistent(commit: &Self::Commit, common: &Self::Common) -> VidResult<()> {
+        <Advz as VidScheme>::is_consistent(commit, common)
+    }
+
+    fn get_payload_byte_len(common: &Self::Common) -> u32 {
+        <Advz as VidScheme>::get_payload_byte_len(common)
+    }
+
+    fn get_num_storage_nodes(common: &Self::Common) -> u32 {
+        <Advz as VidScheme>::get_num_storage_nodes(common)
+    }
+
+    fn get_multiplicity(common: &Self::Common) -> u32 {
+        <Advz as VidScheme>::get_multiplicity(common)
+    }
+}
+
+impl PayloadProver<LargeRangeProofType> for ADVZScheme {
+    fn payload_proof<B>(&self, payload: B, range: Range<usize>) -> VidResult<LargeRangeProofType>
+    where
+        B: AsRef<[u8]>,
+    {
+        self.0
+            .payload_proof(payload, range)
+            .map(LargeRangeProofType)
+    }
+
+    fn payload_verify(
+        &self,
+        stmt: Statement<'_, Self>,
+        proof: &LargeRangeProofType,
+    ) -> VidResult<Result<(), ()>> {
+        self.0.payload_verify(stmt_conversion(stmt), &proof.0)
+    }
+}
+
+impl PayloadProver<SmallRangeProofType> for ADVZScheme {
+    fn payload_proof<B>(&self, payload: B, range: Range<usize>) -> VidResult<SmallRangeProofType>
+    where
+        B: AsRef<[u8]>,
+    {
+        self.0
+            .payload_proof(payload, range)
+            .map(SmallRangeProofType)
+    }
+
+    fn payload_verify(
+        &self,
+        stmt: Statement<'_, Self>,
+        proof: &SmallRangeProofType,
+    ) -> VidResult<Result<(), ()>> {
+        self.0.payload_verify(stmt_conversion(stmt), &proof.0)
+    }
+}
+
+/// Convert a [`VidDisperse<Advz>`] to a [`VidDisperse<VidSchemeType>`].
+///
+/// Foreign type rules prevent us from doing:
+/// - `impl From<VidDisperse<VidSchemeType>> for VidDisperse<Advz>`
+/// - `impl VidDisperse<VidSchemeType> {...}`
+///
+/// and similarly for `Statement`.
+/// Thus, we accomplish type conversion via functions.
+fn vid_disperse_conversion(vid_disperse: VidDisperse<Advz>) -> VidDisperse<ADVZScheme> {
```

Comment:
these feels more like `From<A> for B`?

Reply:
Yes and the comment explains the reason.

Reply:
For `NsTable::ns_range()` comment, I don't know why I cannot reply there.
That's within sequencer scope and I don't want to introduce a circular dependency. So I duplicated the logic.

Reply:
> That's within sequencer scope and I don't want to introduce a circular dependency. So I duplicated the logic.

i see, maybe it's an indication that we should relocate those code to here then.

Reply:
Let's discuss it here: https://github.com/EspressoSystems/espresso-sequencer/issues/2458